GetServicePreferences/handler.ts

```diff
@@ -0,0 +1,191 @@
+
+/**
+ *
+ * @param maybeProfile
+ * @returns
+ */
+const getProfileOrErrorResponse = (
+  maybeProfile: o.Option<RetrievedProfile>
+): te.TaskEither<IResponseErrorNotFound, Profile> =>
+  te.fromEither(
+    e.fromOption(
+      ResponseErrorNotFound(
+        "Profile not found",
+        "The profile you requested was not found in the system."
+      )
+    )(maybeProfile)
+  );
+
+/**
+ * Return a function that returns the service preference for the
+ * given documentId and version, or a default value if not present
+ * The default value depends on the user' settings (mode AUTO/MANUAL)
+ *
+ * @param servicePreferencesModel The service preferences cosmos model
+ * @param fiscalCode the fiscal code
+ * @returns
+ */
+export declare type getUserServicePreferencesT = (params: {
+  readonly documentId: NonEmptyString;
+  readonly mode:
+    | ServicesPreferencesModeEnum.AUTO
+    | ServicesPreferencesModeEnum.MANUAL;
+  readonly version: NonNegativeInteger;
+  readonly fiscalCode: FiscalCode;
+}) => te.TaskEither<IResponseErrorQuery, ServicePreference>;
+const getUserServicePreferencesOrDefault = (
+  servicePreferencesModel: ServicesPreferencesModel
+): getUserServicePreferencesT => ({ fiscalCode, documentId, mode, version }) =>
+  servicePreferencesModel
+    .find([documentId, fiscalCode])
+    .mapLeft(failure =>
+      ResponseErrorQuery(
+        "Error while retrieving the user's service preferences",
+        failure
+      )
+    )
+    .map(maybeServicePref =>
+      maybeServicePref.foldL<ServicePreference>(
+        () => {
+          switch (mode) {
+            case ServicesPreferencesModeEnum.AUTO:
+              return toDefaultEnabledUserServicePreference(version);
+            case ServicesPreferencesModeEnum.MANUAL:
+              return toDefaultDisabledUserServicePreference(version);
+          }
+        },
+        pref => toUserServicePreferenceFromModel(pref)
+      )
+    );
+
+/**
+ * Return a type safe GetServicePreferences handler.
+ */
+export const GetServicePreferencesHandler = (
+  profileModels: ProfileModel,
+  servicePreferencesModel: ServicesPreferencesModel
+): IGetServicePreferencesHandler => {
+  return async (fiscalCode, serviceId) =>
+    profileModels
+      .findLastVersionByModelId([fiscalCode])
+      .mapLeft<
+        IResponseErrorQuery | IResponseErrorNotFound | IResponseErrorConflict
+      >(failure =>
+        ResponseErrorQuery("Error while retrieving the profile", failure)
+      )
+      .chain(getProfileOrErrorResponse)
+      .filterOrElse(
+        nonLegacyServicePreferences,
+        ResponseErrorConflict("Legacy service preferences not allowed")
+      )
+      .chain(profile =>
+        te
+          .fromEither(
+            NonNegativeInteger.decode(
+              profile.servicePreferencesSettings.version
+            )
+          )
+          .mapLeft(_ =>
+            ResponseErrorConflict("Legacy service preferences not allowed")
```

Comment:
We may want to have a different error message, otherwise we cannot be sure which point the handler failed at

Reply:
This one should never happen, since we filter out the `LEGACY`  settings in the previous step.
But yes, I'll change it

Reply:
Done in [b4473e4](https://github.com/pagopa/io-functions-app/pull/170/commits/b4473e49cacf64d97a512a7801b32b9eb8b9db60)